feat(hotlug): trigger hotplug after cloud-init.service

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore

```
<type>(optional scope): <summary>  # no more than 72 characters

A description of what the change being made is and why it is being
made if the summary line is insufficient.  This should be wrapped at
72 characters.

If you need to write multiple paragraphs, feel free.

Fixes GH-NNNNN (GitHub Issue number. Remove line if irrelevant)
LP: #NNNNNN (Launchpad bug number. Remove line if irrelevant)
```
-->
See individual commits.

## Additional Context
<!-- If relevant -->
https://warthogs.atlassian.net/browse/SC-1716

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

```sh
#!/bin/bash

set -eux -o pipefail

make deb

export CLOUD_INIT_CLOUD_INIT_SOURCE=cloud-init_all.deb
export CLOUD_INIT_PLATFORM=ec2
tox -e integration-tests -- tests/integration_tests/modules/test_hotplug.py::test_nics_before_config_trigger_hotplug
```

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

holmanb commented on tests/integration_tests/modules/test_hotplug.py:
```diff
@@ -394,3 +395,145 @@ def test_no_hotplug_triggered_by_docker(client: IntegrationInstance):
+    assert_systemctl_status_code(client, "cloud-final.service", 3)
+
+    added_ip_0 = client.instance.add_network_interface()
+    _wait_till_hotplug_complete(client, expected_runs=1)
```
Before we `_wait_till_...`, I think that we should assert that we are still between `cloud-config.target` done and `cloud-config.service`? If not, then this test didn't do what we thought it did, right?

aciba90 replied:
What I am currently assert is that `cloud-config.target` is finished in line [514: wait_for_cmd(client, "systemctl status cloud-config.target", 0)](https://github.com/canonical/cloud-init/pull/5058/files/6ee132506d6b966e80ee2e2c54639d3bbe6c55af#diff-fe6975e54f120fbe110935bb6e5cd8ba9b5dfe6613eae48787bcbc079c582299R514) and that `cloud-config.service` has not yet started in line  [517: assert_systemctl_status_code(client, "cloud-config.service", 3)](https://github.com/canonical/cloud-init/pull/5058/files/6ee132506d6b966e80ee2e2c54639d3bbe6c55af#diff-fe6975e54f120fbe110935bb6e5cd8ba9b5dfe6613eae48787bcbc079c582299R517). I think that is exactly what you mean, or am I missing something?

holmanb replied:
Those assertions tell us before we add the interface that cloud-config.target is finished and that cloud-config.service has not yet started, but I think that a race is possible. If adding the network interface has some pre-flight check that takes a long time (or later adds something like this on accident), wouldn't it be possible that `add_network_interface()` doesn't actually add the interface until after cloud-config.service has started?

holmanb replied:
With a 6 minute sleep this is all unlikely to happen, but I think that an assertion after adding the nic that we're still in the state we thought we were in would assert that the test is testing what we think it is.

aciba90 replied:
Gotcha, updated.

holmanb commented on tools/hook-hotplug:
```diff
@@ -2,21 +2,31 @@
+# open cloud-init's hotplug-hook fifo rw
+exec 3<>$fifo
+env_params=" \
+--subsystem=${SUBSYSTEM} \
```
nit: Could we indent this and the following lines evaluate to a single string? It looks like you did this to avoid the extra indentation between the args (which I appreciate).  I think we can have the best of both worlds by quoting the strings on each line so that they get concatenated together before assignment.

aciba90 replied:
Do you mean something like:

```sh
env_params=""\
   " --subsystem=${SUBSYSTEM}"\
   " handle"\
   " --devpath=${DEVPATH}"\
   " --udevaction=${ACTION}"
```

This would be nice, but it is not POSIX compatible, it fails in dash and probably on other shells than bash.

I would just KISS and define this var in single line like the following, yes it has 90 characters, but it is more portable and readable IMO.

```sh
env_params=" --subsystem=${SUBSYSTEM} handle --devpath=${DEVPATH} --udevaction=${ACTION}"
```

Or we could use the short version of some parameter if 90 is not acceptable:

```sh
env_params=" -s=${SUBSYSTEM} handle --devpath=${DEVPATH} --udevaction=${ACTION}"
```

holmanb replied:
> This would be nice, but it is not POSIX compatible, it fails in dash and probably on other shells than bash.

Oops, sorry for the bad suggestion.

> I would just KISS and define this var in single line like the following, yes it has 90 characters, but it is more portable and readable IMO.

+1, I'm fine with that

aciba90 replied:
Updated to single line.

holmanb commented on tests/integration_tests/modules/test_hotplug.py:
```diff
@@ -394,3 +395,145 @@ def test_no_hotplug_triggered_by_docker(client: IntegrationInstance):
+    " ability to add_network_interface",
+)
+@pytest.mark.user_data(USER_DATA)
+def test_nics_before_config_trigger_hotplug(client: IntegrationInstance):
```
Nice work getting this test set up!

holmanb reviewed (approved):
@aciba90 Great job on this, and nice test! This looks good to me (besides the comments inline)

I think it'd be good to get a review from another teammate and also probably an SRU reviewer because of the system change, so lets not merge this yet.

holmanb commented:
@TheRealFalcon since you've worked on hotplug before could you please give this another review before we put this in front of SRU reviewer's eyes?

aciba90 commented on tests/integration_tests/modules/test_hotplug.py:
```diff
@@ -394,3 +395,145 @@ def test_no_hotplug_triggered_by_docker(client: IntegrationInstance):
+
+
+def _customize_environment(client: IntegrationInstance):
+    def enable_block_service(service_name: str, content: str):
```
That's a good idea. I did a quick search and didn't find tests pausing cloud-init services. Which ones are you referring to? I'd be happy to refactor / offer this testing ability in a generic manner, but, could we split it as a different issue? And I am also okay with you or anyone else doing it.

holmanb replied:
>> There are some other tests that also do effectively this - this looks reusable. No need to make a change now, but I might refactor this at some point if you don't.
>
> That's a good idea. I did a quick search and didn't find tests pausing cloud-init services.

I was referring to enabling services, but it looks like both of the tests that I was thinking of [[1](https://github.com/canonical/cloud-init/blob/3ad7f287664e41bf6b2a668df78176e62924fc6f/tests/integration_tests/modules/test_ansible.py#L85)][[2](https://github.com/canonical/cloud-init/blob/3ad7f287664e41bf6b2a668df78176e62924fc6f/tests/integration_tests/cmd/test_status.py#L109)] actually do this in userdata. That said, I don't think there's a real advantage to doing it in user-data. It's probably harder to understand and modify. I'm just noting that "install & enable new service and reboot" is a common paradigm in integration tests, so we could create a helper around to simplify things.

> could we split it as a different issue?

Of course - no need to do that in this PR.

aciba90 replied:
Refactored in d935c37954399ec2bf6989d8c18d3f4ab2777ad, thanks.

aciba90 commented on tests/integration_tests/modules/test_hotplug.py:
```diff
@@ -394,3 +395,145 @@ def test_no_hotplug_triggered_by_docker(client: IntegrationInstance):
+    enable_block_service("block-cloud-config", BLOCK_CLOUD_CONFIG)
+    enable_block_service("block-cloud-final", BLOCK_CLOUD_FINAL)
+
+    # Allows ssh access for 1000 user before boot has finished
```
Sure, let me find a pointer.

aciba90 replied:
done

TheRealFalcon reviewed (commented):
Shouldn't we also include a `After=cloud-final.service` in `cloud-init-hotplugd.service`?

With this PR as it is, we will capture all hotplug events that have happened after networking has come up and respond to them immediately. Instead, I thought we wanted to capture those events, wait for the regular cloud-init services to finish, and then respond to the hotplug events. This is due to (theoretical) races with reading and writing the pickle with multiple cloud-init services running simultaneously. 

Maybe this isn't a real problem, in which case this PR would be fine.

TheRealFalcon commented on systemd/cloud-init-hotplugd.service:
```diff
@@ -12,6 +12,7 @@
 [Unit]
 Description=cloud-init hotplug hook daemon
 After=cloud-init-hotplugd.socket
+After=cloud-init.target
```
This should be `cloud-final.servce`, correct?

aciba90 replied:
I think it is equivalent, as `cloud-init.target` is a sync point after `cloud-config.service` and `cloud-final.service`, but more resilient for future changes.

But I am good either way.

TheRealFalcon reviewed (approved):
LGTM!

aciba90 commented:
Since this is touching systemd units, could you, @basak, please take a look?

aciba90 commented:
`main` rebased to fix merge conflicts.

TheRealFalcon commented:
@aciba90 , I think we'll need to actively pursue a reviewer if we want them to take a look at this. Your ping was from before 24.04 release and I'm guessing it has been long lost and forgotten in an inbox.

aciba90 commented:
> @aciba90 , I think we'll need to actively pursue a reviewer if we want them to take a look at this. Your ping was from before 24.04 release and I'm guessing it has been long lost and forgotten in an inbox.

I did so a couple of times in Jira and MM! Re-pinging @basak.

basak commented:
Ack. I was going to try and look at this with Brett this week.

On 7 May 2024 14:32:41 UTC, Alberto Contreras ***@***.***> wrote:
>> @aciba90 , I think we'll need to actively pursue a reviewer if we want them to take a look at this. Your ping was from before 24.04 release and I'm guessing it has been long lost and forgotten in an inbox.
>
>I did so a couple of times in Jira and MM! Re-pinging @basak.
>

holmanb commented:
@aciba90 A couple of concerns:

This is going to enable running a cloud-init service while others are currently running. This carries risk of things that weren't previously possible.

I have some questions:

- What would be the correct behavior if two interface uevents happen simultaneously, or one shortly after the other?

- Is pickling used by the hotplug event? If so, we need to be careful that the primary service cannot be writing to the pickle while the hotplug command is reading and vice versa. This could crash cloud-init.

- Can a socket unit and its paired service unit have different ordering? Since hotplugd.socket depends on cloudinit.service but hotplugd.service depends on cloud-final.service, I haven't found any documentation about what should happen in this configuration.

aciba90 commented:
Thanks for the reviews and questions. I have reworked the PR a bit.

I have modified 62ace19d6796ba4b8bee7db8c583b257b0b409ef and added 57c022d652abf53854546af5abcfffdef7fa3c6f. Please, read the commit messages for full context.

tldr: As architected right now, the `hotplugd.socket` will be active after `cloud-init.service` has finished, enqueueing udev hotplug events, `hotplugd.service` will process the events, one at the time, after `cloud-final.service` has finished. Thus, no races accessing `obj.pkl` or processing udev events will happen.

With this, all the questions are solved, please let me know if you, @holmanb and @basak, have more concerns.

holmanb reviewed (commented):
> and added https://github.com/canonical/cloud-init/commit/57c022d652abf53854546af5abcfffdef7fa3c6f

+1 no objections this matches the other services

> tldr: As architected right now, the hotplugd.socket will be active after cloud-init.service has finished, enqueueing udev hotplug events, hotplugd.service will process the events, one at the time, after cloud-final.service has finished. Thus, no races accessing obj.pkl or processing udev events will happen.

If multiple events are queued, then a read from the pipe will yield:

```
--subsystem=${SUBSYSTEM1} handle --devpath=${DEVPATH1} --udevaction=${ACTION1} --subsystem=${SUBSYSTEM2} handle --devpath=${DEVPATH2} --udevaction=${ACTION2} 
```

Right? It doesn't look like the argparser code will handle this correctly, will it?

As an aside, (and I know that you didn't design it this way) there are several other things that I really don't like about this hotplug code:

- This is the only remaining runtime `bash` dependency that cloud-init has, and it is unjustified - all it does is just read from a file descriptor and run another command.
- The hotplug call chain is unnecessarily convoluted. In `hook-hotplug` we create some cli arguments and then write them to a FIFO. In `cloud-init-hotplugd.service` we have the bash one-liner which concatenates the arguments and calls cloud-init with the subcommands. We could just as easily have passed `SUBSYSTEM`, `DEVPATH` and `ACTION` to the FIFO and in `cloud-init-hotplugd.service` just call `cloud-init devel hotplug-hook` and done the necessary processing in Python. All we would need to do is an `os.read(3, ...)` to read from the fifo that systemd sets up for the socket activated service and then do all of the required processing internal to python code.
- calling something "hotplugd" when it isn't actually a daemon is ugly and confusing. Socket activation is event driven and doesn't involve any daemons - several different people* have been confused by this misnomer. Using socket activation is a significantly better use of resources than a daemon-based implementation would have been.

dermotbradley commented:
> * This is the only remaining runtime `bash` dependency that cloud-init has, and it is unjustified - all it does is just read from a file descriptor and run another command.

Are you referring to tools/hook-hotplug?

I changed this to be POSIX shell rather than Bash-specific in #4772.

holmanb commented:
> > * This is the only remaining runtime `bash` dependency that cloud-init has, and it is unjustified - all it does is just read from a file descriptor and run another command.
> 
> @holmanb Are you referring to tools/hook-hotplug?
> 
> I changed this to be POSIX shell rather than Bash-specific in #4772.

No I was talking about this one:

https://github.com/canonical/cloud-init/blob/00ab32835d25527ff8986fc3cc0ae1c649aa2202/systemd/cloud-init-hotplugd.service#L21

aciba90 commented:
> > and added [57c022d](https://github.com/canonical/cloud-init/commit/57c022d652abf53854546af5abcfffdef7fa3c6f)
> 
> +1 no objections this matches the other services
> 
> > tldr: As architected right now, the hotplugd.socket will be active after cloud-init.service has finished, enqueueing udev hotplug events, hotplugd.service will process the events, one at the time, after cloud-final.service has finished. Thus, no races accessing obj.pkl or processing udev events will happen.
> 
> If multiple events are queued, then a read from the pipe will yield:
> 
> ```
> --subsystem=${SUBSYSTEM1} handle --devpath=${DEVPATH1} --udevaction=${ACTION1} --subsystem=${SUBSYSTEM2} handle --devpath=${DEVPATH2} --udevaction=${ACTION2} 
> ```
> 
> Right? It doesn't look like the argparser code will handle this correctly, will it?

No, that is not right. `hook-hotplug` echoes one line per udev hotplug event to the socket:

```
# contents of /run/cloud-init/hook-hotplug-cmd:
--subsystem=${SUBSYSTEM1} handle --devpath=${DEVPATH1} --udevaction=${ACTION1}
--subsystem=${SUBSYSTEM2} handle --devpath=${DEVPATH2} --udevaction=${ACTION2}
```

Then, systemd will activate a single instance of `cloud-init-hotplugd.service`. 
`cloud-init-hotplugd.service` will use the bash read command which reads only one line, process it and finish.
Then, as the socket has content, systemd  will activate another instance of the service.
And so on.
 
> As an aside, (and I know that you didn't design it this way) there are several other things that I really don't like about this hotplug code:
> 
> * This is the only remaining runtime `bash` dependency that cloud-init has, and it is unjustified - all it does is just read from a file descriptor and run another command.
> 
> * The hotplug call chain is unnecessarily convoluted. In `hook-hotplug` we create some cli arguments and then write them to a FIFO. In `cloud-init-hotplugd.service` we have the bash one-liner which concatenates the arguments and calls cloud-init with the subcommands. We could just as easily have passed `SUBSYSTEM`, `DEVPATH` and `ACTION` to the FIFO and in `cloud-init-hotplugd.service` just call `cloud-init devel hotplug-hook` and done the necessary processing in Python. All we would need to do is an `os.read(3, ...)` to read from the fifo that systemd sets up for the socket activated service and then do all of the required processing internal to python code.

As said above, the use of `read` does ensure only one event is process at the same time. Not sure how `read` behaves in other shell implementations or if it is POSIX complaint, but in any case, I am pretty sure we can implement a *read only one line* POSIX compliant command. But I would vote split that work out of this PR, to limit the scope and because this PR has been open since mid-March.

> * calling something "hotplugd" when it isn't actually a daemon is ugly and confusing. Socket activation is event driven and doesn't involve any daemons - several different people* have been confused by this misnomer. Using socket activation is a significantly better use of resources than a daemon-based implementation would have been.

Daemons can also be activated via sockets: sshd, libvirtd, lvm2-lvmpolld, snapd, uuidd, virtlockd, virtlockd, virtlogd. 

From a user POV, the {udev rule, socket, service} combo acts as a daemon, but yes, it is weird that the *daemon* service is not a long-running process.

I see the value in the renaming, to not break users, we could rename the service and socket and maintain a systemd unit alias pointing to the older names, as a separated ticket too.

Let me know what you think about it, please.

aciba90 commented:
And [57c022d](https://github.com/canonical/cloud-init/commit/57c022d652abf53854546af5abcfffdef7fa3c6f) needs to be patch out for older Ubuntu downstream releases, as those key are not included in older systemd versions.

basak commented:
> tldr: As architected right now, the hotplugd.socket will be active after cloud-init.service has finished, enqueueing udev hotplug events, hotplugd.service will process the events, one at the time, after cloud-final.service has finished. Thus, no races accessing obj.pkl or processing udev events will happen.

What if a hotplug event arrives after cloud-init.service has read the instance metadata, but before it completes and hotplugd.service has started? Wouldn't processing for that event get missed?

basak commented:
> and hotplugd.service has started

nor hotplugd.socket, for that matter.

aciba90 commented:
@basak:

> What if a hotplug event arrives after cloud-init.service has read the instance metadata, but before it completes and hotplugd.service has started? Wouldn't processing for that event get missed? nor hotplugd.socket, for that matter.

In that case, those events are intentionally ignored because the instance could be configured to update the network on [BOOT_LEGACY](https://docs.cloud-init.io/en/latest/explanation/events.html#id1) events, i.e. on every boot: once during `cloud-init-local.service` and another time during `cloud-init.service`, and then if BOOT_LEGACY and HOTPLUG are enabled and if HOTPLUG would run after `cloud-init-local.service`, interfaces could be configured two times.

This is documented in the 62ace19d6796ba4b8bee7db8c583b257b0b409ef's message:

> Note 2: The reason to not extend the scope of this to after init-local is
completed is that during init, network config can be applied if
BOOT_LEGACY is configured as an update event and could race with hotplug.

As this PR is right now, users that want full coverage can enable BOOT_LEGACY and HOTPLUG network update events.

Even if the instance is configured with BOOT_LEGACY and HOTPLUG there is still a small gap when using datasources that determine what NICs to configure via IMDS. Between cloud-init fetches the IMDS at `cloud-init.service` timeframe and `cloud-init.service` finishes where hotplug events will be missed.

I think this gap is okay versus configuring two times the same NIC, and we can think about how to solve that gap in the future if it's seen as a problem.

holmanb reviewed (approved):
> daemon service is not a long-running process.
>
> I see the value in the renaming, to not break users, we could rename the service and socket and maintain a systemd unit alias pointing to the older names, as a separated ticket too.
>
> Let me know what you think about it, please.

- Yes this seems reasonable. Either an alias on old releasrs or maybe we just patch out the renames there. Either way works for me (in a separate PR).

Thanks @aciba90 for your efforts on this! I think that you have addressed my concerns, let's land it.
